From 1d0a10db41a25aa7eb171d0c9c4b64a2b1a0a165 Mon Sep 17 00:00:00 2001
From: Stuart Menefy <stuart.menefy@codasip.com>
Date: Wed, 15 May 2024 18:24:40 +0100
Subject: [PATCH 14/15] hw/riscv/hobgoblin: Add support for proFPGA platform

Currently two versions of the Hobgoblin platform are supported, the
original one using the Genesys 2 hardware, and now the proFPGA. Add an
option to allow the appropriate board to be selected:
  board-type={genesys2|profpga}

Signed-off-by: Stuart Menefy <stuart.menefy@codasip.com>
---
 hw/riscv/hobgoblin.c         | 53 +++++++++++++++++++++++++++++++++++-
 include/hw/riscv/hobgoblin.h |  6 ++++
 2 files changed, 58 insertions(+), 1 deletion(-)

diff --git a/hw/riscv/hobgoblin.c b/hw/riscv/hobgoblin.c
index f54d3cb895..55a3c7de2f 100644
--- a/hw/riscv/hobgoblin.c
+++ b/hw/riscv/hobgoblin.c
@@ -89,6 +89,11 @@ static const memmapEntry_t memmap[] = {
         "riscv.hobgoblin.ram"},
 };
 
+static const memmapEntry_t pro_fpga_memmap[] = {
+    [0] =                  { 0x2000000000, 0x400000000,
+        "riscv.hobgoblin.ram2"},
+};
+
 /* sifive_plic_create() parameters */
 #define HOBGOBLIN_PLIC_NUM_SOURCES      32
 #define HOBGOBLIN_PLIC_NUM_PRIORITIES   7
@@ -372,6 +377,10 @@ static void hobgoblin_machine_init(MachineState *machine)
     /* SRAM exists on FPGA only */
     hobgoblin_add_memory_area(system_memory, &memmap[HOBGOBLIN_SRAM]);
 
+    if (s->board_type == BOARD_TYPE_PROFPGA) {
+        hobgoblin_add_memory_area(system_memory, &pro_fpga_memmap[0]);
+    }
+
     /* add interrupt controller */
     hobgoblin_add_interrupt_controller(s, smp_cpus);
 
@@ -405,9 +414,45 @@ static void hobgoblin_machine_set_boot_from_rom(Object *obj, bool value,
     s->boot_from_rom = value;
 }
 
+static char *hobgoblin_machine_get_board_type(Object *obj, Error **errp)
+{
+    HobgoblinState_t *s = HOBGOBLIN_MACHINE_STATE(obj);
+    const char *result;
+
+    switch (s->board_type) {
+    case BOARD_TYPE_GENESYS2:
+        result = "genesys2";
+        break;
+    case BOARD_TYPE_PROFPGA:
+        result = "profpga";
+        break;
+    default:
+        result = "Unknown";
+        break;
+    }
+
+    return (char*)result;
+}
+
+static void hobgoblin_machine_set_board_type(Object *obj, const char *value,
+                                             Error **errp)
+{
+    HobgoblinState_t *s = HOBGOBLIN_MACHINE_STATE(obj);
+
+    if (!strcmp(value, "genesys2"))
+        s->board_type = BOARD_TYPE_GENESYS2;
+    else if (!strcmp(value, "profpga"))
+        s->board_type = BOARD_TYPE_PROFPGA;
+    else
+        error_setg(errp, "Unrecognised board-type");
+}
+
 static void hobgoblin_machine_instance_init(Object *obj)
 {
-    /* Nothing to be done here. */
+    HobgoblinState_t *s = HOBGOBLIN_MACHINE_STATE(obj);
+
+    s->boot_from_rom = false;
+    s->board_type = BOARD_TYPE_GENESYS2;
 }
 
 static void hobgoblin_machine_class_init(ObjectClass *oc, void *data)
@@ -424,6 +469,12 @@ static void hobgoblin_machine_class_init(ObjectClass *oc, void *data)
                                    hobgoblin_machine_set_boot_from_rom);
     object_class_property_set_description(oc, "boot-from-rom",
         "Load BIOS (default fsbl_rom.xexe) into ROM and boot into it");
+
+    object_class_property_add_str(oc, "board-type",
+                                  hobgoblin_machine_get_board_type,
+                                  hobgoblin_machine_set_board_type);
+    object_class_property_set_description(oc, "board-type",
+        "Set the board type (genesys2 (default) or profpga)");
 }
 
 /* QOM support for HobgoblinState_t */
diff --git a/include/hw/riscv/hobgoblin.h b/include/hw/riscv/hobgoblin.h
index fd504d9619..7feb4db009 100644
--- a/include/hw/riscv/hobgoblin.h
+++ b/include/hw/riscv/hobgoblin.h
@@ -25,11 +25,17 @@
 
 #define MACHINE_TYPE_HOBGOBLIN  MACHINE_TYPE_NAME("hobgoblin")
 
+enum board_type {
+    BOARD_TYPE_GENESYS2,
+    BOARD_TYPE_PROFPGA,
+};
+
 typedef struct {
     /*< private >*/
     MachineState machine; /* QOM: derived from MachineState/TYPE_MACHINE */
     /*< properties >*/
     bool boot_from_rom;
+    enum board_type board_type;
     /*< devices >*/
     RISCVHartArrayState soc;
     DeviceState *plic;
-- 
2.39.3

